ClusterMesh HA

thorn3r opened the pull request:
This adds support for running clustermesh-apiserver deployments with multiple replicas for high availability. Each clustermesh-apiserver pod runs its own etcd cluster. Depending on configuration, either the Cilium Agent or KVStoreMesh instance watches etcd in a remote cluster. All responses from the remote etcd cluster are intercepted and the header is inspected to retrieve the etcd cluster ID. If a failover event occurs and the cluster ID has changed, the remote connection is restarted to ensure that no events are missed and that no invalid data is retained. See individual commit messages for additional details.

```release-note
Add support for deploying clustermesh-apiserver with multiple replicas for high availability.
```

thorn3r committed:
ClusterMesh/helm: support multiple replicas

This commit makes changes to the helm templates for
clustermesh-apiserver to support deploying multiple replicas.

- Use a unique cluster id for etcd:

Each replica of the clustermesh-apiserver deploys its own discrete etcd
cluster. Utilize the K8s downward API to provide the Pod UUID to the
etcd cluster as an initial cluster token, so that each instance has a
unique cluster ID. This is necessary to distinguish connections to
multiple clustermesh-apiserver Pods using the same K8s Service.

- Use session affinity for the clustermesh-apiserver Service

Session affinity ensures that connections from a client are passed to
the same service backend each time. This will allow a Cilium Agent or
KVStoreMesh instance to maintain a connection to the same backend for
both long-living, streaming connections, such as watches on the kv
store, and short, single-response connections, such as checking the
status of a cluster. However, this can be unreliable if the l3/l4
loadbalancer used does not also implement sticky sessions to direct
connections from a particular client to the same cluster node.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
ClusterMesh: validate etcd cluster ID

In a configuration where there are mutliple replicas of the
clustermesh-apiserver, each Pod runs its own etcd instance with a unique
cluster ID. This commit adds a `clusterLock` type, which is a wrapper
around a uint64 that can only be set once. `clusterLock` is used to
create gRPC unary and stream interceptors that are provided to the etcd
client to intercept and validate the cluster ID in the header of all
responses from the etcd server.

If the client receives a response from a different cluster, the
connection is terminated and restarted. This is designed to prevent
accepting responses from another cluster and potentially missing events
or retaining invalid data.

Since the addition of the interceptors allows quick detection of a
failover event, we no longer need to rely on the status checker to
ensure the connection is healthy. Additionally, since service session
affinity can be unreliable, the status checks could trigger a false
failover event and cause a connection restart. To allow creating etcd
clients for ClusterMesh that do not perform status checks, the option
DisableStatusChecks was added to ExtraOptions to skip running the
status checker.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

giorio94 commented on pkg/clustermesh/common/interceptor.go:
```diff
@@ -0,0 +1,118 @@
+		}
+		resp, ok := reply.(etcdResponse)
+		if !ok {
+			cl.errors <- ErrEtcdInvalidResponse
```
Drive-by comment, all these write to channel operations should be guarded by a select statement, to avoid deadlocking after that the first error has been written, given that there would be no receiver at that point.

thorn3r committed:
CI: test ClusterMesh failover in upgrade tests

Update the clustermesh-upgrade CI workflow to test rolling-
upgrades/failover by setting maxSurge=1 and maxUnavailable=0 and adding
a new test using a global service.

The test restarts the clustermesh-apiserver deployment on cluster2 and
deploys a global service to both clusters. Cluster1 should be able to
properly handle the failover event, establish a connection with the
new clustermesh-apiserver pods, and synchronize the endpoints from
cluster2 for the service. To ensure ClusterMesh both with and without
KVStoreMesh enabled can handle failover events, separate upgrading
Cilium and enabling KVStoreMesh for cluster1 into separate steps.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
auto-format ClusterMesh Upgrade workflow yaml

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

giorio94 reviewed (changes requested):
Could you please drop the third commit (the formatting of the workflow). [It already happened once](https://github.com/cilium/cilium/pull/27193/commits/be2b725a90a674ebb498a16155d03e2ec8e8cc1d) (in the other direction), and it caused an incredible amount of issues down the line when backporting subsequent changes.
